theseus/embodied/robot/FK.py

```diff
@@ -0,0 +1,58 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+
+from typing import List, Optional
+
+from theseus.geometry.functional import se3
+from .robot import Robot
+from .link import Link
+from .joint import Joint
+
+# TODO: Add support for joints with DOF>1
+
+
+def ForwardKinematicFactory(robot: Robot, link_names: Optional[List[str]] = None):
+    links: List[Link] = []
+
+    if link_names is None:
+        links = robot.links
+    else:
+        links = [robot.link_map[name] for name in link_names]
+
+    ancestors = []
+
+    for link in links:
+        ancestors += [anc.id for anc in link.ancestors]
+
+    link_ids = sorted(list(set(ancestors)))
```

Comment:
This seems rather inefficient, as the same chain of ancestors would be traversed multiple times if FK for multiple links on the same chain are desired to be computed.
Can we either
1. Do a BFS/DFS from the root, and only keep links that are in the path to links that are in `links`?
2. From each link in `links`, trace back to the root and stop when reaching a link that is previously traced.

Both of these methods can directly output a sorted list of relevant links without repeated traversals.
By doing so we can also completely eliminate the need to cache ancestors.

Also, if for some reason the link ids are not topologically sorted (some child having a smaller id than one of its parents), the methods mentioned above would also cause no problems for computing the FK, in contrary to the current implementation that relies on correct order of link ids.

Reply:
- BFS/DFS is definitely useful and I will refactor the code later with this.
- We already guaranteed that the links are topologically sorted when creating the robot model from URDF, thus, this should not be an issue for our code.